include/seqan3/alphabet/cigar/cigar.hpp

```diff
@@ -0,0 +1,409 @@
+ *       a seqan3::cigar_vector.
+ */
+template<typename length_t = uint32_t>
+//!\cond
+    requires std::UnsignedIntegral<length_t>
+//!\endcond
+class cigar :
+    public cartesian_composition<cigar<length_t>, cigar_op, length_t>
+{
+private:
+    //!\brief The base type.
+    using base_type = cartesian_composition<cigar<length_t>, cigar_op, length_t>;
+
+public:
+
+    //!\brief Equals the char_type of cigar_op type.
+    using char_type = underlying_char_t<cigar_op>;
+
+    //!\brief The template parameter as a type member.
+    using length_type = length_t;
+
+    //!\brief The type combining both types into a streamable string.
+    using string_type = std::string;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    cigar() = default;
+    constexpr cigar(cigar const &) = default;
+    constexpr cigar(cigar &&) = default;
+    constexpr cigar & operator=(cigar const &) = default;
+    constexpr cigar & operator=(cigar &&) = default;
+    ~cigar() = default;
+
+    using base_type::base_type; // Inherit non-default constructors
+
+    using base_type::operator=; // Inherit non-default assignment operators
+
+    //!\copydoc cartesian_composition::cartesian_composition(component_type const alph)
+    SEQAN3_DOXYGEN_ONLY(( constexpr cigar(component_type const alph) {} ))
+    //!\copydoc cartesian_composition::cartesian_composition(indirect_component_type const alph)
+    SEQAN3_DOXYGEN_ONLY(( constexpr cigar(indirect_component_type const alph) {} ))
+    //!\copydoc cartesian_composition::operator=(component_type const alph)
+    SEQAN3_DOXYGEN_ONLY(( constexpr cigar & operator=(component_type const alph) {} ))
+    //!\copydoc cartesian_composition::operator=(indirect_component_type const alph)
+    SEQAN3_DOXYGEN_ONLY(( constexpr cigar & operator=(indirect_component_type const alph) {} ))
+    //!\}
+
+    /*!\name Write functions
+     * \{
+     */
+    //!\brief Assign from a character. This modifies the internal seqan3::cigar_op letter.
+    constexpr cigar & assign_char(char_type const c)
+    {
+        seqan3::assign_char(get<0>(*this), c);
+        return *this;
+    }
+
+    //!\brief Assign from a length value. This modifies the internal length value.
+    constexpr cigar & assign_length(length_type const c)
+    {
+        get<1>(*this) = c;
+        return *this;
+    }
+    //!\}
+
+    /*!\name Read functions
+     * \{
+     */
+    //!\brief Return a character. This reads the internal seqan3::cigar_op letter.
+    constexpr char_type to_char() const noexcept
+    {
+        return seqan3::to_char(get<0>(*this));
+    }
+
+    //!\brief Return the length value. This reads the internal length value.
+    constexpr length_type to_length() const noexcept
+    {
+        return get<1>(*this);
+    }
+
+    //!\brief Return the cigar element as a string. This reads both, the internal seqan3::cigar_op letter and length value.
+    string_type to_string() const noexcept
```

Comment:
Hm, I am not really happy with the general design of this. I would argue that if you define two alphabets where to_char returns e.g. 'M' for the first and "20" for the second, then `to_char` should return "20M" and there shouldn't be a `to_string`. That's just super confusing. A string is always multiple letters of one alphabet and here we convert a single letter into multiple ones...

I think this can only be properly solved by adding a "small_string" class to our container that has variable size <= 16 and uses a fixed 16byte array internally, similar to the small-string-optimisation part of std::string. We could make it inherit from view_base so that views accept it as constant-time copyable.

Reply:
Probably you could use fixed size constexpr_string as a basis